cmd/restic/cmd_cleanup_packs.go

```diff
@@ -0,0 +1,330 @@
+						Length: blob.Length,
+					})
+			}
+		}
+
+		// TODO: Parallelize repacking
+		for id := range repackPacks {
+			if !opts.DryRun {
+				blobs := blobsPerPack[id]
+				debug.Log("processing pack %v, blobs: %v", id, len(blobs))
+				err := repack(ctx, repo, blobs, nil)
+				if err != nil {
+					return err
+				}
+			} else {
+				if !gopts.JSON && gopts.verbosity >= 2 {
+					Verbosef("would have repacked pack %v.\n", id.Str())
+				}
+			}
+
+			// Also remove repacked pack at the end!
+			removePacks.Insert(id)
+		}
+
+		err = repo.Flush(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(repackPacks) > 0 && !opts.DryRun {
+		Verbosef("updating index files...\n")
+
+		// TODO: This is a hack to get the index entries of repacked blobs
+		// Should be replaced if there is another index implementation
+		notFinalIdx := (repo.Index()).(*repository.MasterIndex).NotFinalIndexes()
+		if len(notFinalIdx) != 1 {
+			return errors.Fatal("should only have one unfinalized index!!")
+		}
+		repackIndex := notFinalIdx[0]
+
+		err = ChangePacksInIndex(opts, gopts, repo, repackIndex)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(removePacks) != 0 {
+		Verbosef("deleting packs (unused and repacked)\n")
+		fileHandles := make(chan restic.Handle)
+		go func() {
+			for packID := range removePacks {
+				fileHandles <- restic.Handle{Type: restic.DataFile, Name: packID.String()}
+			}
+			close(fileHandles)
+		}()
+		DeleteFiles(gopts, opts.DryRun, repo, len(removePacks), fileHandles)
+
+	}
+
+	Verbosef("done\n")
+	return nil
+}
+
+const numDeleteWorkers = 8
+
+func DeleteFiles(gopts GlobalOptions, dryrun bool, repo restic.Repository, totalCount int, fileHandles <-chan restic.Handle) {
+	var wg sync.WaitGroup
+	bar := newProgressMax(!gopts.Quiet, uint64(totalCount), "files deleted")
+	bar.Start()
+	wg.Add(numDeleteWorkers)
+	for i := 0; i < numDeleteWorkers; i++ {
+		go func() {
+			for h := range fileHandles {
+				if !dryrun {
+					err := repo.Backend().Remove(gopts.ctx, h)
+					if err != nil {
+						Warnf("unable to remove file %v from the repository\n", h.Name)
+					}
+					if !gopts.JSON && gopts.verbosity >= 2 {
+						Verbosef("%v was removed.\n", h.Name)
+					}
+				} else {
+					if !gopts.JSON && gopts.verbosity >= 2 {
+						Verbosef("would have removed %v.\n", h.Name)
+					}
+				}
+				bar.Report(restic.Stat{Blobs: 1})
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	bar.Done()
+}
+
+func repack(ctx context.Context, repo restic.Repository, blobs []restic.Blob, p *restic.Progress) (err error) {
+
+	var buf []byte
+	for _, entry := range blobs {
+		// if blob is not in index, don't write it
+		if !repo.Index().Has(entry.ID, entry.Type) {
+			debug.Log("not writing blob %v (not in index)", entry.ID)
+			continue
+		}
+
+		buf = buf[:]
+		if uint(len(buf)) < entry.Length {
+			buf = make([]byte, entry.Length)
+		}
+		buf = buf[:entry.Length]
+
+		// TODO: Use encrypted blob and just save it instead decrypting and encrypting again
```

Comment:
I'll need to have a closer look on that..

Reply:
I will change the code to use `repository/repack.Repack` instead.
However, having a first look at this `Repack` function it seems to me that this function has the same problem if the pack file is corrupted, or did I miss something?

Reply:
I've added stricter checks for blob correctness in #2674. The deduplication currently works by marking all packs with duplicate blobs for rewriting. A rewritten blob is removed from `keepBlobs`. Right now `Repack` can't properly handle damaged packs. With some additional checks, it would be relatively simple to pick a correct copy by skipping the broken copies. If no valid copy is found, then the packs should be kept. That will probably require some additional checks to prevent endless additional copies of blobs from partially broken packs.